### ISSUE TITLE
5.2.0 Now requires running InModuleScope Scripts independently

### ISSUE THREAD

**Issue by Christoba** [pester/Pester#1939]
## General summary of the issue

* Our test scripts align to a general format the employs `InModuleScope` for wrapping the `Describe`.
* Our test scripts are 1-1 with functions that are either "public" (exported) and "private" to a module.

With 5.1.1, scripts could all be tested with a single run of `Invoke-Pester`

With 5.2.0, scripts run successfully when run independently `Invoke-Pester -Path` but will fail when run in a single invocation of `Invoke-Pester` with what appear to an absence of mocks and or vars declared within `BeforeAll` blocks.

Suspicion:
The "isolation" introduced/implemented in 5.2.0 is impacted or is impacting InModuleScope wrapping.

Being able to test modules and all functions within via a _single_ `InModuleScope` is very useful as
* it avoids the need of dot sourcing private functions in each script
* it avoids the need to annotate each function mock with the module to which it belongs

## Describe your environment

```powershell
Pester version     : 5.2.0 <redacted>\PowerShell\Modules\pester\5.2.0\Pester.psm1
PowerShell version : 7.1.3
OS version         : Microsoft Windows NT 10.0.19042.0
```

## Steps to reproduce

Example script format. Note that BeforeAll is also tried from within the Describe.

```powershell
[Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSUseDeclaredVarsMoreThanAssignments", "")]
param()

$moduleRoot = (Split-Path -Parent $PSScriptRoot)
Remove-Module "MyModule" -Force -ErrorAction Ignore
Import-Module "$moduleRoot\MyModule.psd1" -Force

InModuleScope MyModule {
  BeforeAll {
    $GlobalValue = 'Some Test Value'
    Mock Run-ThirdPartyFunction
  }

  Describe 'Run-PrivateFunction' {

    Context 'Some Context' {

      It 'Should have a GlobalValue but does not' {
        # Local mocks
        Mock Run-AnotherPrivateFunction

        $result = Run-PrivateFunction
        $result.Value | Should -Be $GlobalValue
      }

     It 'Should have Run-ThirdPartyFunction mocked but does not' {
        # Local mocks
        Mock Run-AnotherPrivateFunction

        $result = Run-PrivateFunctionCallingRunThirdPartyFunction
      }
    }
  }
}
```

## Expected Behavior

`Invoke-Pester` on multiple script files isolates the invocations of each script (share nothing).

## Current Behavior

`Invoke-Pester`:
Assertion failures due to absence of Vars/Mocks set.

## Workaround

`Get-Childitem *.Tests.ps1 -recurse | ForEach-Object { Invoke-Pester -CI -Path $_.FullName }`

**Comment by nohwnd** [comment]
Could you show the errors please? 

I am trying to get a repro, this is what I am trying 

[isolation.zip](https://github.com/pester/Pester/files/6464420/isolation.zip)

```powershell
# myModule.psm1

function Run-ThirdPartyFunction () { throw "don't call me" }
function Run-PrivateFunction () { throw "don't call me either" }


Export-ModuleMember -Function ""

```
```powershell
# ex.tests.ps1 - just used a different path to module and used psm1

InModuleScope MyModule {
  BeforeAll {
    $GlobalValue = 'Some Test Value'
    Mock Run-ThirdPartyFunction
  }

  Describe 'Run-PrivateFunction' {

    Context 'Some Context' {

      It 'Should have a GlobalValue but does not' {
        # Local mocks
        Mock Run-AnotherPrivateFunction

        $result = Run-PrivateFunction
        $result.Value | Should -Be $GlobalValue
      }

     It 'Should have Run-ThirdPartyFunction mocked but does not' {
        # Local mocks
        Mock Run-AnotherPrivateFunction

        $result = Run-PrivateFunctionCallingRunThirdPartyFunction
      }
    }
  }
}
```

```
Discovery finished in 522ms.
Running tests.
[-] Run-PrivateFunction.Some Context.Should have a GlobalValue but does not 23ms (22ms|1ms)
 CommandNotFoundException: Could not find Command Run-AnotherPrivateFunction

[-] Run-PrivateFunction.Some Context.Should have Run-ThirdPartyFunction mocked but does not 22ms (21ms|1ms)
 CommandNotFoundException: Could not find Command Run-AnotherPrivateFunction

Tests completed in 751ms
Tests Passed: 0, Failed: 2, Skipped: 0 NotRun: 0
```
I see this, which I think is caused by no longer looking in the current scope when function is not found in module (when inside of module we don't find the function because it is not listed in get-module, and I remember removing that code, because it never made sense to me and it did not break anything).  Now I know why it might be there, and we need to improve when it runs to check the module name.

But from your description you are able to run script when running one by one, I can't.

**Comment by Christoba** [comment]
[PesterSample.zip](https://github.com/pester/Pester/files/6467878/PesterSample.zip)

No problem - I've reproduced (at least the mocking issue) in the attached module. Not currently seeing the variable issue in this trimmed repro but that could be more subtle.

Pester 5.2.0:
```powershell
import-module pester -requiredversion 5.2.0
invoke-pester

Starting discovery in 2 files.
WARNING: The names of some imported commands from the module 'MyModule' include unapproved verbs that might make them less discoverable. To find the commands with unapproved verbs, run the Import-Module command again with the Verbose parameter. For a list of approved verbs, type Get-Verb.
WARNING: The names of some imported commands from the module 'MyModule' include unapproved verbs that might make them less discoverable. To find the commands with unapproved verbs, run the Import-Module command again with the Verbose parameter. For a list of approved verbs, type Get-Verb.
Discovery finished in 157ms.
Running tests.
[-] Run-PrivateFunction.Private Function Context.Should have Run-AnotherPrivateFunction mocked but does not 53ms (52ms|1ms)
 Expected Run-AnotherPrivateFunction in module MyModule to be called at least 1 times but was called 0 times
 at Should -Invoke Run-AnotherPrivateFunction -Scope It, C:\git2\PesterSample\src\MyModule\Private\Run-PrivateFunction.Tests.ps1:27
 at <ScriptBlock>, C:\git2\PesterSample\src\MyModule\Private\Run-PrivateFunction.Tests.ps1:27
[+] C:\git2\PesterSample\src\MyModule\Public\Run-PublicFunction.Tests.ps1 324ms (101ms|201ms)
Tests completed in 1.09s
Tests Passed: 3, Failed: 1, Skipped: 0 NotRun: 0
invoke-pester C:\git2\PesterSample\src\MyModule\Private\Run-PrivateFunction.Tests.ps1

Starting discovery in 1 files.
WARNING: The names of some imported commands from the module 'MyModule' include unapproved verbs that might make them less discoverable. To find the commands with unapproved verbs, run the Import-Module command again with the Verbose parameter. For a list of approved verbs, type Get-Verb.
Discovery finished in 26ms.
Running tests.
[+] C:\git2\PesterSample\src\MyModule\Private\Run-PrivateFunction.Tests.ps1 397ms (27ms|347ms)
Tests completed in 401ms
Tests Passed: 2, Failed: 0, Skipped: 0 NotRun: 0
```

Pester 5.1.1:
```powershell
import-module pester -requiredversion 5.1.1
invoke-pester

Starting discovery in 2 files.
WARNING: The names of some imported commands from the module 'MyModule' include unapproved verbs that might make them less discoverable. To find the commands with unapproved verbs, run the Import-Module command again with the Verbose parameter. For a list of approved verbs, type Get-Verb.
WARNING: The names of some imported commands from the module 'MyModule' include unapproved verbs that might make them less discoverable. To find the commands with unapproved verbs, run the Import-Module command again with the Verbose parameter. For a list of approved verbs, type Get-Verb.
Discovery finished in 168ms.
[+] C:\git2\PesterSample\src\MyModule\Private\Run-PrivateFunction.Tests.ps1 758ms (192ms|432ms)
[+] C:\git2\PesterSample\src\MyModule\Public\Run-PublicFunction.Tests.ps1 236ms (51ms|163ms)
Tests completed in 1s
Tests Passed: 4, Failed: 0, Skipped: 0 NotRun: 0
```

**Comment by nohwnd** [comment]
The repro is still a bit broken but I see the difference when invoking the file by itself and when running them all. The problem here is the Import-Module -Force on the top of the file and our resolve.

When you run Discovery you will import MyModule and all the scriptblocks will attach to it because you use InModuleScope. Then you progress to the next file and you import MyModule again. It has the same name and code, but from powershell point of view it is entirely different module - different session state. Your scriptblocks will attach to this MyModule (let's call it MyModule-B), and everything is happy. 

Discovery is done and you start the Run phase, and your tests start executing. 

You enter BeforeAll and you define a mock. Here is where the problem comes. We resolve modules by name, and don't look into the current session state. You are running in MyModule-A, but there is also MyModule-B imported. Pester does Get-Module MyModule and gets MyModule-B, it defines mock there, and continues.

Then you invoke the function that should be mocked and it is not. Your mocks were defined correctly, but in a totally different session state.

If you have more files this would happen over and over again, until you get to a file that is the last one to force import the module. This file is lucky, it has two references to MyModule-B, it both runs in it's session state, and will see it by Get-Module. The same thing happens, but because both session state and Get-Module will resolve to the same module MyModule-B, the Mocks will work as expected. You also see the same thing when running the files one by one.

Simple repro:
```powershell
Invoke-Pester -Container (New-PesterContainer -ScriptBlock {
    
    Get-Module m | Remove-Module
    New-Module -Name m -ScriptBlock {
        $id = [System.Guid]::NewGuid().Guid
        function Get-Id { $id }

        Export-ModuleMember -Function ""
    } | Import-Module
    
    
    InModuleScope m {
        Describe 'm1' {
            It 'we are in m, but also have m imported and they are not the same' {
            
                $mA = Get-Id
                $mB = &(Get-Module m) { Get-Id }
                $mA | Should -Not -Be $mB 
            }

            It 'i1' {
                Mock Test-Path { "mock" }
                Test-Path "aaa" | Should -Be "mock"
            }
        }
    }

    # importing a differen module with the same name
    # (when reproing with scriptblock module you must not 
    # use the same scriptblock, otherwise it won't work
    # the same way as with files, probably because files are 
    # parsed again every time )
    Get-Module m | Remove-Module
    New-Module -Name m -ScriptBlock {
        $id = [System.Guid]::NewGuid().Guid
        function Get-Id { $id }

        Export-ModuleMember -Function ""
    } | Import-Module

    InModuleScope m {
        Describe 'm2' {
            It 'i2' {
                Mock Test-Path { "mock 2" }
                Test-Path "aaa" | Should -Be "mock 2"
            }
        }
    }
})
```
```
[-] m1.i1 1.03s (1.03s|1ms)
 Expected 'mock', but got $false.
 at Test-Path "aaa" | Should -Be "mock", C:\Users\jajares\Desktop\aaa.ps1:27
 at <ScriptBlock>, C:\Users\jajares\Desktop\aaa.ps1:27

[DBG]: PS C:\Users\jajares\Downloads\PesterSample\src\MyModule> 
Tests completed in 3.8s
Tests Passed: 2, Failed: 1, Skipped: 0 NotRun: 0
```

I need to fix the way the command is resolved.

**Comment by Christoba** [comment]
That explanation makes sense. A secondary workaround then is to no longer force import the module in order to maintain a single module for Pester to find.

Tweaking the import before the InModuleScope looks to work for the resolution issue and gets us back to a single Pester run.

```powershell
[Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSUseDeclaredVarsMoreThanAssignments", "")]
param()

$moduleRoot = (Split-Path -Parent $PSScriptRoot)
if (-not (Get-Module MyModule)) { Import-Module "$moduleRoot\MyModule.psd1" }

InModuleScope MyModule {
  BeforeAll {
    $GlobalValue = 'This is a test value.'
    Mock Run-AnotherPrivateFunction {}
    Mock Test-Path {}
  }

  Describe 'Run-PrivateFunction' {

    Context 'Private Function Context' {
      It 'Should have a GlobalValue but does not' {
        Mock Run-AnotherPrivateFunction {}
        $result = Run-PrivateFunction
        $result | Should -Be $GlobalValue
      }

      It 'Should have Run-AnotherPrivateFunction mocked but does not' {

        $result = Run-PrivateFunction
        Should -Invoke Run-AnotherPrivateFunction -Scope It
      }
    }
  }
}
```

**Comment by nohwnd** [comment]
This is fixed in 5.2.1 by #1941
